osrd-schemas: add script to truncate an infrastructure

Castavo opened the pull request.

Castavo committed:
osrd-schemas: add script to truncate an infrastructure

Signed-off-by: Baptiste Prevot <pro.baptiste.prevot@gmail.com>

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+    def keep_track_section(track_section: TrackSection) -> bool:
+        return bbox.intersects(LineString(track_section.geo.coordinates))
+
+    new_infra.track_sections = list(filter(keep_track_section, infra.track_sections))
```
Shouldn't you use `filter_track_sections` instead (unused so far)? And remove `keep_track_section`.

woshilapin replied:
I removed the function in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/03e809515eb2b7e3f92e50dfe6f42f09e0158af2 and actually inlined also `keep_track_section` in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/25871f49c76fd6a82d43c6087f5130bb26eef864. Tell me if that’s ok with you.

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+        "bbox",
+        help="Bounding box as GeoJSON polygon coordinates\n"
+        'For example "[[0, 0], [0, 1], [1, 1], [1, 0]]"\n'
+        "A way to get it is to use https://geojson.io/",
```
Maybe a bit nitpicky, but a "real" bounding box is only made of 2 coords, for the same example it would be `[[0,0],[1,1]]`.
As the feature is finer, I'd change the example at least, maybe even rename the param:
```suggestion
        "bounds",
        help="Boundaries as GeoJSON polygon coordinates\n"
        'For example ♦️: "[[0, 0], [1, 1], [2, 0], [1, -1]]"\n'
        "A way to get it is to use https://geojson.io/",
```

woshilapin replied:
I’ve slightly diverged from your comment since I used “boundary” instead of “bound” but I’ve made the change in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/ff205ec667afbad1f16fee15e325cc5196e87e43.

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+        raise ValueError(f"Could not find the connected track for {current_track_id}")
+    return connected_track
+
+
```
Not mandatory: is it possible to document the existence of this somewhere? ✅
Not sure about where I'd go search if looking for this, though...

Also documenting the behavior of the filter would be nice at least in `--help` (filter tracks, then keep all related to tracks).

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+    new_infra = truncate_infra(infra, bbox)
+
+    print("Writing new infra")
+    args.output.write_text(new_infra.model_dump_json(exclude_none=True))
```
Not mandatory: adding tests would be nice.

Thinking of:
* "easy" global test (no change of infra)
* "easy" empty test (small shape outside or in the middle of infra)
* "didactic" test on the fact that track-section intersection is the root of filtering?

Maybe based on small_infra? (See hand-tests in a later [comment](https://github.com/OpenRailAssociation/osrd/pull/13526#issuecomment-3389033296))

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+        )
+        if new_track_ranges:
+            new_announcement_ranges = filter_track_ranges(
+                neutral_section.announcement_track_ranges, track_section_ids
```
Not sure: could it lead to inconsistent neutral sections (missing announcement)?
It may not be worth the hassle to requalify track-sections, though...

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+            filtered_switches.append(switch)
+        elif n_ports_included > 1:
+            print(
+                f"Switch {switch.id} should be included in the bbox, but isn't handled yet"
```
Not sure it's a good idea: we could also keep the original switches and add fake (or the same id but shortened?) 0-length track-sections + buffer-stop.
That way everything would be handled in the same way and we would keep the ids.
But 0-length tracks could end-up in buggy code later in OSRD?

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+                res.append(route)
+            else:
+                # This is an edge case not yet handled
+                print(f"Warning: dropping {route.id} that is partially included")
```
How about handling all the same way:
```py
if n_switches_included > 0 or entry_included or exit_included:
    # iterate over [entry + switches + exit]:
    #    create route(id=route.id + "_n")
    #    (drop "_n" if route complete)
```

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+
+        if entry_included and exit_included:
+            if n_switches_included == len(route.switches_directions):
+                # The route is fully included, no need to change anything
```
Isn't there a (not so) corner-case with the `LINK` case?
IIRC, links are not listed anymore in routes, no?
We could end-up with a track-section excluded from the new_infra and linked to this new_infra by 2 links: the retained route would be broken for core I guess.

bougue-pe reviewed (approved):
I like the feature, the code is clear and not so tricky.
Even dealing with corner cases looks doable at a decent price... I'd like to have it available in editoast+front :heart:

I noted all the ideas, I'm not sure about the workforce we want to put in it, so I let you decide.
Adding TODOs may sometimes be enough for now, too?

bougue-pe commented:
I did small_infra tests:
* ✅ big shape around -> the same as input (except some `null` fields)
* ✅ small shape completely away -> an empty infra
* ❌ the only partial shape I tried -> errors on pydantic: ✅ Corrected
```sh
❯ uv run scripts/truncate_infra.py ../../tests/data/infras/small_infra/infra.json infra.json "[                                                                                                         [
              -0.14969977501792187,
              49.54358601203904
            ],
            [
              -0.3917723787158707,
              49.483236027590976
            ],
            [
              -0.39779075836571565,
              49.45411986754067
            ],
            [
              -0.18982230601724837,
              49.461509188307474
            ],
            [
              -0.14969977501792187,
              49.54358601203904
            ]
          ]"
Reading infra
Truncating infra
Keeping 13/31 track sections
Traceback (most recent call last):
  File "/home/pierretienne/dev/sources/osrd/python/osrd_schemas/scripts/truncate_infra.py", line 375, in <module>
    new_infra = truncate_infra(infra, bbox)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pierretienne/dev/sources/osrd/python/osrd_schemas/scripts/truncate_infra.py", line 78, in truncate_infra
    new_infra.routes = filter_routes(
                       ^^^^^^^^^^^^^^
  File "/home/pierretienne/dev/sources/osrd/python/osrd_schemas/scripts/truncate_infra.py", line 280, in filter_routes
    route.entry_point = BufferStopReference(id=new_bufferstop_id)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pierretienne/dev/sources/osrd/python/osrd_schemas/.venv/lib/python3.12/site-packages/pydantic/main.py", line 253, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for BufferStopReference
type
  Field required [type=missing, input_value={'id': 'buffer_stop_PA2'}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.11/v/missing
```

EDIT: fixable by changing `BufferStopReference.type` param to:
`type: Literal["BufferStop"] = "BufferStop"`
Then I get correct pathfinding when loading it, no core warning when creating a train, no error on infra ✅.

woshilapin commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+                Direction.START_TO_STOP
+                if the_port.endpoint == Endpoint.BEGIN
+                else Direction.STOP_TO_START,
+            )
```
nitpick: To avoid indexing in the list something already appended, you could do this


```suggestion
            new_buffer_stop = BufferStop(
                id=f"buffer_stop_{switch.id}",
                track=the_port.track,
                position=0
                if the_port.endpoint == Endpoint.BEGIN
                else track_length_by_track[the_port.track],
            )
            former_switch_id_to_buffer_stop_id[switch.id] = (
                new_buffer_stop.id,
                Direction.START_TO_STOP
                if the_port.endpoint == Endpoint.BEGIN
                else Direction.STOP_TO_START,
            )
            new_buffer_stops.append(new_buffer_stop)
```

woshilapin replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/bddc5efbd130f7e85dc7d9d5f400270b7febf6e0.

woshilapin commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+                # This is an edge case not yet handled
+                print(f"Warning: dropping {route.id} that is partially included")
+
+        elif entry_included or exit_included:
```
nitpick: it’s technically a XOR (the OR does work because you already treated the AND above, but yet, if code moves, making this one a XOR right now might be better).

```suggestion
        elif entry_included ^ exit_included:
```

woshilapin replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/f1fed4994d927d88a530c3b059ee8e59a8528e63.

woshilapin commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+    return res
+
+
+def filter_track_ranges(
```
nitpick: perhaps this function can be inlined like it already is [for some others](https://github.com/OpenRailAssociation/osrd/pull/13526/files#diff-2aaecf07016d05ac94dfe637f6de08a8e5cbe2bf7c7519605c2b4463dbfa3d1cR46-R54). Also, sometimes `list(filter())` is used, sometimes comprehension lists are used, but I’d say this might be a good thing to always use the same technic for filtering (I have no opinion on which, comprehension lists are more idiomatic maybe, but `list(filter())` makes it clearer that the intention is to “filter”)

woshilapin replied:
I’m resolving this comment since this function is reused multiple times so it does help to factorize into a function instead of inlining it.

woshilapin replied:
Used comprehension list every where in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/25871f49c76fd6a82d43c6087f5130bb26eef864

woshilapin commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,378 @@
+        res.append(switch_dir_on_track[0])
+        switch_id, dir = res[-1]
+
+        switch_directions.remove(res[-1])
```
nitpick: Isn’t it more straightforward to understand when we use directly the object instead of the indirection of “the last element of an array in which we just inserted an element”?


```suggestion
        switch_id, dir = switch_dir_on_track[0]

        switch_directions.remove(switch_dir_on_track[0])
```

woshilapin replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/e2bfeae553e0f297c3b037aa9d351fbd67b3fe2d.

woshilapin reviewed (approved):
Looks like a useful tool, even with all the unhandled cases (I’m sure people who need this tool, and reach these edge cases will fix it). I’ve mostly made some readability suggestions.

Castavo commented:
Regarding the unhandled cases, they can be circumvented by providing a GeoJson outline which cuts track sections in a simple way (i.e. mostly convex shape)

I made this script as a quick way to get the needed functionnality, and meant to commit it in here without much "maintainability" effort such as tests, because it was not meant for regular/production use and more as a utility to pick up and update if needed.

I think that you both reviewed this scrit as a fully-fledged feature, and in this sense, I agree with all your comments.

However I don't have enough time to work on and update this PR that much. Please feel free to take the PR over and push it to feature-like quality, if you have the motivation and/or time !

woshilapin committed:
osrd-schemas: use a intermediate variable for readibility

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
osrd-schemas: use a XOR instead of an OR

The AND is already addressed above, so this condition is only triggered
in the case of XOR. If code moves around later, having a XOR right now
will avoid to run that piece of code in the wrong condition.

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
osrd-schemas: use comprehension lists everywhere

The goal is to have some kind of homogeneity in the code instead of a
mix between comprehension list and `list(filter())`. Comprehension list
seems more idiomatic. Actually, it was plan to remove `filter`, `reduce`
and friend from Python when comprehension lists were introduced.

https://web.archive.org/web/20250921103937/https://www.artima.com/weblogs/viewpost.jsp?thread=98196

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
osrd-schemas: remove unused function

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
osrd-schemas: change the wording from ‘bbox’ to ‘boundary’

A ‘bbox’ is usually only 2 coordinates. In this case, it’s more about a
list of coordinates to describe a boundary.

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
osrd-schemas: use variable instead of backward reference in an array

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin commented:
@bougue-pe I pushed a fix for the problem you (and I also) had with the creation of `BufferStopReference` in https://github.com/OpenRailAssociation/osrd/pull/13526/commits/fbfc6ecc60e0d50bc33c188dd8d145fbd0896b4f.

woshilapin committed:
osrd-schemas: add documentation about ‘truncate-infra.py’

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin commented:
I addressed all the comments I had and also some of the obvious one that @bougue-pe had. I’ve also added a small piece in the `README.md` for lack of a better idea about where it should live (see https://github.com/OpenRailAssociation/osrd/pull/13526/commits/5800cf957d4f5189d4a89da8838dc40b7f03ee26). I’ve tested with success on the example mentioned above as well as some other more involved infrastucture of France with success.

I do not intend to work more on it, I’ll leave it in your hands @bougue-pe. To me, it is mergeable as-is if we don’t want to spend more time on it.

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,376 @@
+}
+
+
+def truncate_infra(infra: RailJsonInfra, bbox: Polygon) -> RailJsonInfra:
```
Ditto for all `bbox`:
```suggestion
def truncate_infra(infra: RailJsonInfra, boundary: Polygon) -> RailJsonInfra:
```

woshilapin replied:
I’ve updated the commit https://github.com/OpenRailAssociation/osrd/pull/13526/commits/c23fa520ccec5e3316fd069c99640b5aefb2d3d4. I’ve also rebased the PR.

bougue-pe commented on python/osrd_schemas/scripts/truncate_infra.py:
```diff
@@ -0,0 +1,376 @@
+            ]
+            if entry_included:
+                route.exit_point = BufferStopReference(
+                    id=new_bufferstop_id, type="BufferStop"
```
I'd still do the add of default in osrd_schemas, though.

woshilapin replied:
I’ve read your first comment too fast! 🤦 But I do agree. See https://github.com/OpenRailAssociation/osrd/pull/14406.

woshilapin replied:
#14406 has been merged, so I removed the `type = "BufferStop"` when needed. The script still work, as expected.

bougue-pe committed:
fixup! osrd-schemas: add script to truncate an infrastructure

Signed-off-by: Pierre-Etienne Bougué <bougue.pe@proton.me>
